connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufCatalystDataConversionSuite.scala

```diff
@@ -0,0 +1,212 @@
+
+class ProtobufCatalystDataConversionSuite
+    extends SparkFunSuite
+    with SharedSparkSession
+    with ExpressionEvalHelper {
+
+  private def checkResult(
+      data: Literal,
+      descFilePath: String,
+      messageName: String,
+      expected: Any): Unit = {
+    checkEvaluation(
+      ProtobufDataToCatalyst(
+        CatalystDataToProtobuf(data, descFilePath, messageName),
+        descFilePath,
+        messageName,
+        Map.empty),
+      prepareExpectedResult(expected))
+  }
+
+  protected def checkUnsupportedRead(
+      data: Literal,
+      descFilePath: String,
+      actualSchema: String,
+      badSchema: String): Unit = {
+
+    val binary = CatalystDataToProtobuf(data, descFilePath, actualSchema)
+
+    intercept[Exception] {
+      ProtobufDataToCatalyst(binary, descFilePath, badSchema, Map("mode" -> "FAILFAST")).eval()
+    }
+
+    val expected = {
+      val expectedSchema = ProtobufUtils.buildDescriptor(descFilePath, badSchema)
+      SchemaConverters.toSqlType(expectedSchema).dataType match {
+        case st: StructType =>
+          Row.fromSeq((0 until st.length).map { _ =>
+            null
+          })
+        case _ => null
+      }
+    }
+
+    checkEvaluation(
+      ProtobufDataToCatalyst(binary, descFilePath, badSchema, Map("mode" -> "PERMISSIVE")),
+      expected)
+  }
+
+  protected def prepareExpectedResult(expected: Any): Any = expected match {
+    // Spark byte and short both map to Protobuf int
+    case b: Byte => b.toInt
+    case s: Short => s.toInt
+    case row: GenericInternalRow => InternalRow.fromSeq(row.values.map(prepareExpectedResult))
+    case array: GenericArrayData => new GenericArrayData(array.array.map(prepareExpectedResult))
+    case map: MapData =>
+      val keys = new GenericArrayData(
+        map.keyArray().asInstanceOf[GenericArrayData].array.map(prepareExpectedResult))
+      val values = new GenericArrayData(
+        map.valueArray().asInstanceOf[GenericArrayData].array.map(prepareExpectedResult))
+      new ArrayBasedMapData(keys, values)
+    case other => other
+  }
+
+  private val testingTypes = Seq(
+    StructType(StructField("int32_type", IntegerType, nullable = true) :: Nil),
+    StructType(StructField("double_type", DoubleType, nullable = true) :: Nil),
+    StructType(StructField("float_type", FloatType, nullable = true) :: Nil),
+    StructType(StructField("bytes_type", BinaryType, nullable = true) :: Nil),
+    StructType(StructField("string_type", StringType, nullable = true) :: Nil))
+
+  private val catalystTypesToProtoMessages: Map[DataType, String] = Map(
+    IntegerType -> "IntegerMsg",
+    DoubleType -> "DoubleMsg",
+    FloatType -> "FloatMsg",
+    BinaryType -> "BytesMsg",
+    StringType -> "StringMsg")
+
+  testingTypes.foreach { dt =>
+    val seed = 1 + scala.util.Random.nextInt((1024 - 1) + 1)
+    val filePath = testFile("protobuf/catalyst_types.desc").replace("file:/", "/")
+    test(s"single $dt with seed $seed") {
```

Comment:
@SandishKumarHN Is it a flaky test?
<img width="1057" alt="image" src="https://user-images.githubusercontent.com/5399861/197108901-f3a9b36f-0fc7-47c7-a4a1-e04162d66c65.png">

Reply:
@wangyum yeah it randomly fails, the fix is done here https://github.com/apache/spark/pull/38286

Reply:
@rangadi the above error occurs randomly, Can we contact someone from the spark-connect team to help us fix the https://github.com/apache/spark/pull/38286 error?